### DISCUSSION TITLE
Looking for suggestions on how to set up CFWheels for two sub-apps under a parent website

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1322] [Q&A] [answered]
New to CFWheels and still learning. I see how CFWheels sets up the directories for a single site/app.  But I'm needing help with how to set up the CFWheels directories when there are a couple of sub-apps and some common shared code.  Here is how I'll diagram the relationships between objects:
```
/ (the parent-app webroot)
/welcome.cfm (contains links for going to app1 or app2 or some shared areas)
/app1/ (sub-app folder)
  section1
  section2
/app2/ (sub-app folder)
  section1
  section2
/attachments (shared folder)
/notes (shared folder)
/library (shared folder)
```

I'm envisioning a separate CFWheels directory structure under each app (app1 and app2) where each can have a section1 page and a section2 page, etc. But then it's not clear how the code outside of app1 and app2 would be included in CFW.  But if I put one CFW directory list under / like is usually what is done, then how can you have separate controllers and models and views under each app (app1 and app2)? 
Is this even possible?

**Comment by dbelanger** [comment]
Hey Bruce, the minute you add Wheels to app1 and app2, you'll have two separate apps with web roots that are app1 and app2 respectively.  Basically, to access anything outside of those webroot, just add mappings to your config/app.cfm.  In your example, `/app1/config/app.cfm` may have:
```
this.mappings['/attachments'] = GetDirectoryFromPath( GetBaseTemplatePath() ) & "../attachments";
this.mappings['/notes'] = GetDirectoryFromPath( GetBaseTemplatePath() ) & "../notes";
this.mappings['/library'] = GetDirectoryFromPath( GetBaseTemplatePath() ) & "../library";
```

**Reply by cfcoder** [reply]
Thanks @dbelanger. This project starts out with a login page and then a dashboard landing page if successful login. From their, there are tabs to select which app to go into.  Given the tabbed UI, the user may switch back and forth between the two apps in a given session. My preference is to not have multiple wheels trees and keep this one app but to have a file hierarchy that allows controllers in separate directories but also maybe some controllers that are shared between the apps.  Perhaps the file structure might be like:
```
/ (webroot)
/controllers
  /app1
    /controller1.cfc
    /controller2.cfc
  /app2
    /controller1.cfc
    /controller2.cfc
  /shared_controller1.cfc
/views
  /app1
    /template1.cfm
    /template2.cfm
  /app2
    /template1.cfm
    /template2.cfm
  /shared_template.cfm

```
but I don't know how to do the routing or if this can work. I think Namespaces may hold the key?

I'm envisioning having url paths like /app1/controller1 or /app2/controller2, etc.

**Comment by dbelanger** [comment]
I should mention that moving Wheels outside of the web root is currently in the works and once that's ready, you won't need to repeat your Wheels directory in your sub apps.  You'll have something like this:
```
/webroot (the parent-app webroot)
  /welcome.cfm (contains links for going to app1 or app2 or some shared areas)
  /app1/ (sub-app folder)
    section1
    section2
  /app2/ (sub-app folder)
    section1
    section2
/attachments (shared folder)
/notes (shared folder)
/library (shared folder)
/wheels (shared folder)
```
and your Application.cfc for each app will simply say:
```
component output="false" {
	include "../../wheels/functions.cfm";
}
```
This will effectively keep the framework code and all of your shared folders out of your webroot

**Comment by bpamiri** [comment, marked as answer]
I also wanted to mention that Wheels has a more advance feature called Namespaces that allows you to partition your single application into several "sub folders" or URL patterns. In this mode, you would have a single Wheels app but you could create two name spaces for /app1 and /app2. This way you could have /app1/home/index and /app2/home/index being separate files but be able to share models and code between the two.

It really depends on how isolated you need to make /app1 and /app2.

https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/routing#namespaces

**Reply by cfcoder** [reply]
I think this sounds like what I need to use.  Is having the "home" in the path necessary or is that just your example to illustrate? i.e. would home be a controller?

I think (as I explained to Paul above) I'm seeking a kind of hybrid where some controllers are just for app1 and some for app2 and some for either.  Same for views, where perhaps a dialog is shared or a table view or something.

If there are some example projects publicly available that do something like this, it would be most helpful!

**Reply by bpamiri** [reply]
That is just an example. Your hierarchy can be anything you can dream up.

**Reply by zainforbjs** [reply]
@cfcoder I will go with @bpamiri recommendations. we can mark as answer for this discussion and close.
